Enhancement/financial request menu UI

Herlin6 opened the pull request.

Herlin6 committed:
add: icon for the financial request status

Herlin6 committed:
edit: datetime format in Created At tooltip

Herlin6 committed:
edit: filter in financial request menu

- edit search bar and SortBy filter
- add trigger for a modal popup in the search bar
- add SortBy button in table header
- add clear filter button

fanesz commented on src/renderer/src/common/components/Input/SearchBar/index.tsx:
```diff
@@ -23,6 +24,10 @@ const SearchBar = (props: Props) => {
     onChange(debouncedValue);
   }, [debouncedValue]);
 
+  useEffect(() => {
+    setTempValue(value);
+  }, [value]);
+
```
kan value sudah di passing lewat onChange(), ini untuk??

fanesz commented on src/renderer/src/common/components/Input/SearchBar/index.tsx:
```diff
@@ -32,7 +37,7 @@ const SearchBar = (props: Props) => {
       placeholder={placeholder}
       sx={{
         "& .MuiInputBase-input": {
-          height: "1.2rem"
+          height: "0.9rem"
```
knp di pendekin?

fanesz commented on src/renderer/src/pages/Financial/List/FinancialTable.tsx:
```diff
@@ -1,26 +1,45 @@
+    getFilterLabel
+  } = useFinancialFilters();
   const { handleClickRow } = useFinancialDetails();
+  const filterSortBy = filters.find((filter) => filter.key === "sort");
+  const { handleClearSortBy } = useFinancialFilters();
+  const [searchValue, setSearchValue] = useState<string>("");
+  const handleChangeSearchBar = useCallback((value: string) => {
+    setSearchValue(value);
+    handleChangeSearch(value);
+  }, []);
+  const handleClear = useCallback(() => {
+    setSearchValue("");
+    handleClearSortBy();
+  }, [handleClearSortBy]);
```
searchValue kan sudah di handling di component searchbar, disitu pake debouncer, jadi pas input, 0.5 detik, auto enter.
jangan pake usestate/usecallback, tapi seluruh ini di bikin function, trus pindahin ke custom hooks useFinancialFilters.
yang diclear itu state dari context, jadi useState itu unnecessary di design pattern ini.

untuk clear jg masih ngebug, misal filter by status, kadang statusnya tetep approved.

fanesz commented on src/renderer/src/pages/Financial/context/index.tsx:
```diff
@@ -106,3 +110,24 @@ const FinancialProvider: FC<{ children: ReactNode }> = ({ children }) => {
 export { FinancialProvider, useFinancialContext };
 export type { StateType };
+
+export function StatusIcon(status: string): JSX.Element {
+  let icon;
+  const commonIconStyle = {
+    fontSize: "medium",
+    opacity: 0.5
+  };
+  if (status === "Rejected") {
+    icon = <ErrorOutlineIcon color="error" sx={commonIconStyle} />;
+  } else if (status === "Pending") {
+    icon = <WarningAmberIcon color="warning" sx={commonIconStyle} />;
+  } else {
+    icon = <CheckCircleOutlineIcon color="success" sx={commonIconStyle} />;
+  }
+
+  return (
+    <>
+      {icon} {status}
+    </>
+  );
+}
```
???????????????????????????????????????

kalo emang butuh komponen terpisah, bikin lah di folder partials, context itu cuma untuk simpen STATE / DATA

fanesz commented on src/renderer/src/pages/Financial/List/hooks/useFinancialList.ts:
```diff
@@ -12,12 +13,12 @@ const useFinancialList = (): HookReturn => {
-    { label: "status", width: "w-24" },
-    { label: "note", width: "w-52" },
-    { label: "date", width: "w-28" }
+    { label: "id", width: "w-20", sortBy: "" },
+    { label: "username", width: "w-44", sortBy: "" },
+    { label: "amount", width: "w-32", sortBy: "" },
+    { label: "status", width: "w-28", sortBy: sortBy(2) },
+    { label: "note", width: "w-52", sortBy: "" },
+    { label: "date", width: "w-24", sortBy: sortBy(1) }
```
sort ttp harus di context, maenin function untuk set yang di sort, bukan disimpen jadi variable object

fanesz commented on src/renderer/src/pages/Financial/List/hooks/useFinancialFilters.ts:
```diff
@@ -27,15 +31,27 @@ const useFinancialFilters = (): HookReturn => {
+      ],
+      value: ["asc", "desc"]
+    },
+    {
+      key: "status",
+      label: "Status",
+      options: [
+        { label: "Approved", value: "1" },
+        { label: "Rejected", value: "2" },
+        { label: "Pending", value: "3" }
+      ],
+      value: ["1", "2", "3", ""]
     }
```
status ini kan sudah di fetch oleh fetchFilterOptions.
kalo kayak ini, misal ada status baru, harus diubah, ini masih hardcode.
status itu di fetch ke endpoint /v1/enums/payment-status karena enums table data

fanesz replied:
jadi state ini cuma untuk tampung options, bukan value, value ditampung di context

fanesz commented on src/renderer/src/pages/Financial/List/FinancialTable.tsx:
```diff
@@ -38,38 +57,51 @@ const FinancialTable = () => {
             <td>{row.username}</td>
             <td>{rupiahFormatter(row.amount)}</td>
-            <td>{row.paymentStatus}</td>
+            <td>{StatusIcon(row.paymentStatus)}</td>
```
status icon terlalu gelep, gap antar logo dan text masih terlalu jauh

fanesz replied:
mending pindahin disisi kanan 
misal:
Approved ✓
Rejected x

fanesz commented on src/renderer/src/pages/Financial/List/FinancialTable.tsx:
```diff
@@ -1,26 +1,45 @@
   TableBody,
-  TableFilter,
-  TableFilterItem,
+  // TableFilter,
+  // TableFilterItem,
```
ilangin comment

fanesz commented on src/renderer/src/pages/Financial/Details/FinancialDetailsContent.tsx:
```diff
@@ -29,7 +29,10 @@ const FinancialDetailsContent = (props: Props) => {
           <div className="-mt-1">
-            <Tooltip title={currentFinreq.created_at} arrow>
+            <Tooltip
+              title={moment(currentFinreq.created_at).format("dddd, hh:mm A")}
```
tampilin detik

fanesz commented on src/renderer/src/common/types/index.ts:
```diff
@@ -31,6 +31,7 @@ export type FilterType = {
   key: string;
   label: string;
   options: CommonOptions[];
+  value: Array<string>;
```
```suggestion
```
di context, sudah ada variable `filters`, value ditampung disitu, bukan useState ini

fanesz commented on src/renderer/src/pages/Financial/List/hooks/useFinancialFilters.ts:
```diff
@@ -83,6 +100,47 @@ const useFinancialFilters = (): HookReturn => {
+  }
+  const getFilterLabel = filterLabel;
+
+  const handleClearSortBy = () => {
+    handleChangeSortBy("order_by", ["asc"]);
+    handleChangeSortBy("status", []);
+  };
```
ini yg di clear itu context

fanesz replied:
setState
...prevState
filters: {
... copas filters default yg skrg di context
}

fanesz commented on src/renderer/src/common/components/Table/TableSortBy/index.tsx:
```diff
@@ -0,0 +1,21 @@
+    handleChangeSortBy(firstFilter.key, firstFilter.value);
+  }
+
+  return (
+    <>
+      <SwapVertIcon
+        sx={{
+          fontSize: "medium"
+        }}
+        onClick={handleSortItem}
+      />
+    </>
+  );
+}
```
```suggestion
  return (
    <SwapVertIcon
      sx={{
        fontSize: "medium"
      }}
      onClick={handleSortItem}
    />
  );
}
```
uneccessary

fanesz reviewed (changes requested):
komponen searchbar jangan di otak atik, itu sudah bener, delay 0.5s bakal auto apply, itulah fitur debouncer. kalo mau ditambahin fitur _skip debounce on enter_ baru boleh.

header yg punya fitur sort blm dibikin pointer cursornya

fitur clear filter masih banyak bug

fitur sortby berdasarkan ascending/descending hilang...

Herlin6 committed:
edit: financial request menu

- add StatusIcon function in partials
- remove useless type declaration
- edit icon in financial request status
- add seconds to datetime format in Created At tooltip
- edit handleChangeSortBy
 and handleClearSortBy functions in useFinancialRequest
